### ISSUE TITLE
Argument 'instance' (type interface) may not be null

### ISSUE THREAD

**Issue by A-Metwally** [zhanghai/gnome-shell-extension-transparent-top-bar#7]
**Update**:
 `journalctl` output from **Debian 11**, gnome-shell v3.36.4
```
Aug 05 22:17:53 hostname systemd-logind[553]: Lid closed.
Aug 05 22:17:53 hostname systemd-logind[553]: Suspending...
Aug 05 22:17:53 hostname gnome-shell[1752]: Object .MetaWindowActorX11 (0x55f42c2dfd20), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Aug 05 22:17:53 hostname gnome-shell[1752]: == Stack trace for context 0x55f4299b84f0 ==
Aug 05 22:17:53 hostname gnome-shell[1752]: #0   7fff8ced1900 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:571 (1d7bb6c40 @ 25)
Aug 05 22:17:53 hostname gnome-shell[1752]: #1   7fff8ced1a30 b   /home/username/.local/share/gnome-shell/extensions/transparent-top-bar@zhanghai.me/extension.js:46 (dafe5b961f0 @ 586)
Aug 05 22:17:53 hostname gnome-shell[1752]: #2   7fff8ced2640 b   resource:///org/gnome/shell/ui/extensionSystem.js:107 (3a7bc3968358 @ 394)
Aug 05 22:17:53 hostname gnome-shell[1752]: #3   7fff8ced3100 b   resource:///org/gnome/shell/ui/extensionSystem.js:615 (3a7bc396a5b0 @ 15)
Aug 05 22:17:53 hostname gnome-shell[1752]: #4   7fff8ced3b50 b   self-hosted:266 (1d7bace20 @ 259)
Aug 05 22:17:53 hostname gnome-shell[1752]: #5   55f42c125710 i   resource:///org/gnome/shell/ui/extensionSystem.js:614 (3a7bc396a538 @ 98)
Aug 05 22:17:53 hostname gnome-shell[1752]: #6   55f42c125680 i   resource:///org/gnome/shell/ui/extensionSystem.js:632 (3a7bc396a628 @ 82)
Aug 05 22:17:53 hostname gnome-shell[1752]: #7   7fff8ced4a10 b   self-hosted:1009 (3d67df99e6a0 @ 423)
Aug 05 22:17:53 hostname gnome-shell[1752]: #8   7fff8ced4b10 b   resource:///org/gnome/gjs/modules/core/_signals.js:133 (1d7bac880 @ 427)
Aug 05 22:17:53 hostname gnome-shell[1752]: #9   7fff8ced5630 b   resource:///org/gnome/shell/ui/sessionMode.js:195 (3d67df924f88 @ 286)
Aug 05 22:17:53 hostname gnome-shell[1752]: #10   55f42c125520 i   resource:///org/gnome/shell/ui/sessionMode.js:156 (3d67df924da8 @ 40)
Aug 05 22:17:53 hostname gnome-shell[1752]: #11   55f42c125468 i   resource:///org/gnome/shell/ui/screenShield.js:571 (3d67df90ff88 @ 191)
Aug 05 22:17:53 hostname gnome-shell[1752]: #12   55f42c1253a8 i   resource:///org/gnome/shell/ui/screenShield.js:620 (3d67df910088 @ 419)
Aug 05 22:17:53 hostname gnome-shell[1752]: #13   55f42c125310 i   resource:///org/gnome/shell/ui/screenShield.js:219 (3d67df90f268 @ 65)
Aug 05 22:17:53 hostname gnome-shell[1752]: #14   7fff8ced6310 b   self-hosted:1011 (3d67df99e6a0 @ 454)
Aug 05 22:17:53 hostname gnome-shell[1752]: #15   7fff8ced6410 b   resource:///org/gnome/gjs/modules/core/_signals.js:133 (1d7bac880 @ 427)
Aug 05 22:17:53 hostname gnome-shell[1752]: #16   55f42c125258 i   resource:///org/gnome/shell/misc/loginManager.js:198 (3a7bc39063d0 @ 202)
Aug 05 22:17:53 hostname gnome-shell[1752]: #17   7fff8ced73e0 b   self-hosted:1013 (3d67df99e6a0 @ 492)
Aug 05 22:17:53 hostname gnome-shell[1752]: #18   7fff8ced8a40 b   resource:///org/gnome/gjs/modules/core/_signals.js:133 (1d7bac880 @ 427)
Aug 05 22:17:53 hostname gnome-shell[1752]: #19   55f42c1250c8 i   resource:///org/gnome/gjs/modules/core/overrides/Gio.js:169 (1d7b92880 @ 39)
Aug 05 22:17:53 hostname gnome-shell[1752]: JS ERROR: Extension transparent-top-bar@zhanghai.me: Error: Argument 'instance' (type interface) may not be null
                                             _init/GObject.Object.prototype.disconnect@resource:///org/gnome/gjs/modules/core/overrides/GObject.js:571:24
                                             disable@/home/username/.local/share/gnome-shell/extensions/transparent-top-bar@zhanghai.me/extension.js:46:27
                                             _callExtensionDisable@resource:///org/gnome/shell/ui/extensionSystem.js:107:32
                                             _disableAllExtensions/<@resource:///org/gnome/shell/ui/extensionSystem.js:615:22
                                             _disableAllExtensions@resource:///org/gnome/shell/ui/extensionSystem.js:614:52
                                             _sessionUpdated@resource:///org/gnome/shell/ui/extensionSystem.js:632:18
                                             _emit@resource:///org/gnome/gjs/modules/core/_signals.js:133:47
                                             _sync@resource:///org/gnome/shell/ui/sessionMode.js:195:14
                                             pushMode@resource:///org/gnome/shell/ui/sessionMode.js:156:14
                                             activate@resource:///org/gnome/shell/ui/screenShield.js:571:34
                                             lock@resource:///org/gnome/shell/ui/screenShield.js:620:14
                                             _prepareForSleep@resource:///org/gnome/shell/ui/screenShield.js:219:22
                                             _emit@resource:///org/gnome/gjs/modules/core/_signals.js:133:47
                                             _prepareForSleep@resource:///org/gnome/shell/misc/loginManager.js:198:14
                                             _emit@resource:///org/gnome/gjs/modules/core/_signals.js:133:47
                                             _convertToNativeSignal@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:169:19
Aug 05 22:17:55 hostname systemd[1]: Reached target Sleep.
```

**Comment by A-Hussein-M** [comment]
System info: Fedora 32, shell version 3.36.2, Wayland session

**Comment by zhanghai** [comment]
How can I reproduce this? I'm running gnome-shell 3.36.2 on Arch Linux, and I cannot reproduce it. The line the stacktrace pointed to also looks innocent to me.

**Comment by A-Hussein-M** [comment]
can't really reproduce it since the crashes are so random, but might be related to some programs running in full-screen. I noticed the crash for the first time after waking the laptop from suspend, the only two apps running were nautilus and WPS office viewing a presentation in full screen. last occurrence was yesterday when i put the laptop into suspend while playing a full-screen youtube video in firefox.

**Comment by zhanghai** [comment]
I tried starting `wpp`, presenting an empty slide, lock screen and suspend, then press power button and unlock, but still couldn't reproduce this. How often do you see it?

**Comment by A-Hussein-M** [comment]
it's so random which is the annoying part about it, I tried suspending the laptop or locking the screen in every single scenario with several apps (including `WPS` and `Firefox`) playing in full screen but still couldn't reproduce.
I close the laptop's lid more than 20 times a day so the first day I noticed the crash, it crashed nearly every time I closed the lid and I had to re-login several times that day, since then it crashes once or twice a day, or doesn't crash at all, last time It crashed was two days ago.

**Comment by mythmakerseven** [comment]
I've got the same issue on a Debian laptop and Ubuntu desktop. There also seems to be no identifiable pattern here, it just randomly crashes about once per day. It seems to occur when the display shuts off, whether for sleep or otherwise (my desktop doesn't sleep, but the screen turns off for inactivity).

Ubuntu 20.04 / gnome-shell 3.36.3
Debian Sid / 3.36.4
both running X11

````
Aug 23 20:46:03 hostname gnome-shell[1970]: Reenabling 'Window Is Ready' Notification
Aug 23 20:46:03 hostname gnome-shell[1970]: Ignoring excess values in shadow definition
Aug 23 20:46:03 hostname gnome-shell[1970]: Object .MetaWindowActorX11 (0x559b1e5d7570), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Aug 23 20:46:03 hostname gnome-shell[1970]: == Stack trace for context 0x559b1ce30700 ==
Aug 23 20:46:03 hostname gnome-shell[1970]: #0   7ffdcef1ef10 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:571 (f7f2a5b6c40 @ 25)
Aug 23 20:46:03 hostname gnome-shell[1970]: #1   7ffdcef1f040 b   /home/username/.local/share/gnome-shell/extensions/transparent-top-bar@zhanghai.me/extension.js:46 (1c65156a9718 @ 586)
Aug 23 20:46:03 hostname gnome-shell[1970]: #2   7ffdcef1fd20 b   resource:///org/gnome/shell/ui/extensionSystem.js:108 (28e260764628 @ 394)
Aug 23 20:46:03 hostname gnome-shell[1970]: #3   559b26482f68 i   resource:///org/gnome/shell/ui/extensionSystem.js:624 (28e26076a8f8 @ 15)
Aug 23 20:46:03 hostname gnome-shell[1970]: #4   7ffdcef20c20 b   self-hosted:266 (28e260738448 @ 259)
Aug 23 20:46:03 hostname gnome-shell[1970]: #5   559b26482ed8 i   resource:///org/gnome/shell/ui/extensionSystem.js:623 (28e26076a880 @ 98)
Aug 23 20:46:03 hostname gnome-shell[1970]: #6   559b26482e48 i   resource:///org/gnome/shell/ui/extensionSystem.js:641 (28e26076a970 @ 82)
Aug 23 20:46:03 hostname gnome-shell[1970]: #7   7ffdcef21ae0 b   self-hosted:1009 (1bdc0ba23088 @ 423)
Aug 23 20:46:03 hostname gnome-shell[1970]: #8   7ffdcef21be0 b   resource:///org/gnome/gjs/modules/core/_signals.js:133 (1c65156e8970 @ 427)
Aug 23 20:46:03 hostname gnome-shell[1970]: #9   7ffdcef228c0 b   resource:///org/gnome/shell/ui/sessionMode.js:198 (1bdc0ba26880 @ 286)
Aug 23 20:46:03 hostname gnome-shell[1970]: #10   559b26482be8 i   resource:///org/gnome/shell/ui/sessionMode.js:159 (1bdc0ba266a0 @ 40)
Aug 23 20:46:03 hostname gnome-shell[1970]: #11   559b26482b30 i   resource:///org/gnome/shell/ui/screenShield.js:574 (1bdc0ba13880 @ 191)
Aug 23 20:46:03 hostname gnome-shell[1970]: #12   559b26482a98 i   resource:///org/gnome/shell/ui/screenShield.js:311 (1bdc0ba0ed30 @ 36)
Aug 23 20:46:03 hostname gnome-shell[1970]: #13   559b264829c8 i   self-hosted:1011 (1bdc0ba23088 @ 454)
Aug 23 20:46:03 hostname gnome-shell[1970]: #14   559b26482938 i   resource:///org/gnome/shell/ui/lightbox.js:195 (28e260733d30 @ 29)
Aug 23 20:46:03 hostname gnome-shell[1970]: #15   559b264828a8 i   resource:///org/gnome/shell/ui/environment.js:75 (f7f2a58d5b0 @ 98)
Aug 23 20:46:03 hostname gnome-shell[1970]: #16   559b26482818 i   resource:///org/gnome/shell/ui/environment.js:149 (f7f2a58d9e8 @ 14)
Aug 23 20:46:03 hostname gnome-shell[1970]: JS ERROR: Extension transparent-top-bar@zhanghai.me: Error: Argument 'instance' (type interface) may not be null
                                           _init/GObject.Object.prototype.disconnect@resource:///org/gnome/gjs/modules/core/overrides/GObject.js:571:24
                                           disable@/home/username/.local/share/gnome-shell/extensions/transparent-top-bar@zhanghai.me/extension.js:46:27
                                           _callExtensionDisable@resource:///org/gnome/shell/ui/extensionSystem.js:108:32
                                           _disableAllExtensions/<@resource:///org/gnome/shell/ui/extensionSystem.js:624:22
                                           _disableAllExtensions@resource:///org/gnome/shell/ui/extensionSystem.js:623:52
                                           _sessionUpdated@resource:///org/gnome/shell/ui/extensionSystem.js:641:18
                                           _emit@resource:///org/gnome/gjs/modules/core/_signals.js:133:47
                                           _sync@resource:///org/gnome/shell/ui/sessionMode.js:198:14
                                           pushMode@resource:///org/gnome/shell/ui/sessionMode.js:159:14
                                           activate@resource:///org/gnome/shell/ui/screenShield.js:574:34
                                           _onLongLightbox@resource:///org/gnome/shell/ui/screenShield.js:311:18
                                           onComplete@resource:///org/gnome/shell/ui/lightbox.js:195:18
                                           _makeEaseCallback/<@resource:///org/gnome/shell/ui/environment.js:75:13
                                           _easeActor/<@resource:///org/gnome/shell/ui/environment.js:149:56
````
That first thing in the log ("Reenabling 'Window Is Ready' Notification") is from the NoAnnoyance extension (https://github.com/BjoernDaase/noannoyance). I've uninstalled it to see if it's interfering with Transparent Top Bar, but it's unlikely to be the root cause because that extension is not installed on my laptop, which is also affected by this issue.

**Comment by aldonogueira** [comment]
Since the bug occurs at the line with `actor.disconnect(signalId);` I tried to surround it with a try-catch like this:
```javascript
...
for (const signalId of signalIds) {
    try {
        actor.disconnect(signalId);
    } 
    catch (error) {
        log('[Transparent Top Bar] Encountered an error cleaning up extension \n'
          + 'actor: ' + typeof(actor) + ' = ' + actor + '\n'
          + 'signalId: ' + typeof(signalId) + ' = ' + signalId + '\n'
          + error);
    }
}
...
```
I hoped this would bring more information on what kind of object throws the error, but the result was another exception:
```
JS ERROR: Extension transparent-top-bar@zhanghai.me: Error: Argument 'actor' (type interface) may not be null
init/Clutter.Actor.prototype.toString@resource:///org/gnome/shell/ui/environment.js:281:19
disable@/home/aldo/.local/share/gnome-shell/extensions/transparent-top-bar@zhanghai.me/extension.js:51:41
_callExtensionDisable@resource:///org/gnome/shell/ui/extensionSystem.js:108:32
_disableAllExtensions/<@resource:///org/gnome/shell/ui/extensionSystem.js:624:22
_disableAllExtensions@resource:///org/gnome/shell/ui/extensionSystem.js:623:52
_sessionUpdated@resource:///org/gnome/shell/ui/extensionSystem.js:641:18
_emit@resource:///org/gnome/gjs/modules/core/_signals.js:133:47
_sync@resource:///org/gnome/shell/ui/sessionMode.js:198:14
pushMode@resource:///org/gnome/shell/ui/sessionMode.js:159:14
activate@resource:///org/gnome/shell/ui/screenShield.js:574:34
_onLongLightbox@resource:///org/gnome/shell/ui/screenShield.js:311:18
onComplete@resource:///org/gnome/shell/ui/lightbox.js:195:18
_makeEaseCallback/<@resource:///org/gnome/shell/ui/environment.js:75:13
_easeActor/<@resource:///org/gnome/shell/ui/environment.js:149:56
```
Which is interesting because it now tell us the actor is this Clutter thing. I hope it helps  

I'm using Ubuntu 20.04 upgraded from 19.10 and 19.04 before that. The theme is the gnome default. Other active extensions are:
- PulseAudio_Loopback_Device@jocagovi.gmail.com
- sound-output-device-chooser@kgshank.net

And from from ubuntu repository:
- gnome-shell-extension-caffeine
- gnome-shell-extension-system-monitor
- gnome-shell-extension-ubuntu-dock
- gnome-shell-extension-appindicator

**Comment by zhanghai** [comment]
@aldonogueira Could you try removing the ` + ' = ' + actor` part? I suppose this is the code causing `Clutter.Actor.prototype.toString()` to fail, and removing it allows us to know which actor is misbehaving (e.g. `Main.overview`, `Main.sessionMode` etc)

**Comment by zhanghai** [comment]
Nevermind, the object is a `MetaWindowActorX11` according to https://github.com/zhanghai/gnome-shell-extension-transparent-top-bar/issues/7#issue-630213828 and https://github.com/zhanghai/gnome-shell-extension-transparent-top-bar/issues/7#issuecomment-678952710 .

> Object .MetaWindowActorX11 (0x55f42c2dfd20), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.

It seems that the C object was somehow previously GC'ed and when we call disconnect in GJS, the GJS runtime finds out and throws us an exception in [
`ObjectInstance::check_gobject_disposed()](https://gitlab.gnome.org/GNOME/gjs/-/blob/gnome-3-36/gi/object.cpp#L197).

I looked into mutter/clutter source code a lot but still couldn't find a path where a `ClutterContainer` or `ClutterGroup` would have a child destroyed without the `actor-removed` signal, because the child always removes itself from the parent during destroy. Maybe the suggestion in the log that some C code is destroying it with the vfuncs, but I couldn't find them in `MetaWindowActorX11` or `MetaWindowActor` anyway.

Could you reproduce this on 3.38.0? I'm running 3.38.0 now and haven't notice the issue, maybe it's fixed in the new version?

**Comment by zhanghai** [comment]
And in case you can still reproduce it, could you try b67d420? This is the best fix I can think of, and if it still doesn't fix the issue, we can only resort to dirtry try-catch for this specific issue.

**Comment by A-Hussein-M** [comment]
> Could you reproduce this on 3.38.0?

Not on my system, I have been running version 3.38 since ubuntu 20.10 reached beta (for about a month now).
I'll keep you updated if that changes.

**Comment by aldonogueira** [comment]
I don't have a gnome 3.38 here, but I think it would be important to find a fix or workaround for 3.36 since so many people stick with LTS Ubuntu.  

I didn't know this ALT-F2 + R thing. I little error appeared in the logs, but the top bar is still transparent:
```
JS ERROR: TypeError: this._windowSignalIds.get(...) is undefined
_onWindowActorRemoved@/home/aldo/.local/share/gnome-shell/extensions/transparent-top-bar@zhanghai.me/extension.js:82:54
_initializeUI/<@resource:///org/gnome/shell/ui/main.js:236:16
```
The line number is different because of my try-catch. It's the first line in _onWindowActorRemoved() function, a for loop.  

I will try the fix you suggested before
